src/gleam/regex.gleam

```diff
@@ -0,0 +1,122 @@
+//// This module contains regular expression matching functions for strings.
+//// The matching algorithms of the library are based on the PCRE library, but not
+//// all of the PCRE library is interfaced and some parts of the library go beyond
+//// what PCRE offers. Currently PCRE version 8.40 (release date 2017-01-11) is used.
+
+import gleam/list
+import gleam/option.{Option}
+
+pub external type Regex
+
+/// The details about a particular match:
+///
+/// - match — the full string of the match.
+/// - index — the index of the match in the original string.
+/// - number — each match is numbered, starting from 1
+/// - submatches — a Regex can have subpatterns, sup-parts that are in parentheses.
+///
+pub type Match {
+  Match(
+    match: String,
+    index: Int,
+    number: Int,
+    submatches: List(Option(String)),
+  )
+  Error(Nil)
```

Comment:
I would prefer not to have this Error record here, instead return `Result(Match, Nil)` so that it can be used with `try` and the `result` module

Reply:
I removed this Error record. Took some inspiration from: https://github.com/gleam-lang/stdlib/blob/597b1e93fc2e0ea36bb443fa393c4bb8f0e1600d/src/gleam/atom.gleam#L20